File: src/algorithms/sortandsearch/SortColors.java
package algorithms.sortandsearch; 

import java.util.*; 

public class SortColors {
    /* Function: Sort red(0), white(1), blue(2) so that they group together. 
     * Solution: Two pointer quick sort. Notice when p2 == p1, cannot swap. 
     */ 
    
    public void sortColors(int[] A) {
        if (A == null || A.length == 0) {
            return; 
        }
        
        int red = -1, white = -1; 
        for (int i = 0; i < A.length; ++i) {
            if (A[i] == 0) {
                swap(A, ++red, i); 
                ++white; 
                if (white > red) {
                    swap(A, white, i); 
                }
            } else if (A[i] == 1) {
                swap(A, ++white, i); 
            }
        }
        
        return; 
    }
    
    private void swap(int[] A, int p1, int p2) {
        int tmp = A[p1]; 
        A[p1] = A[p2]; 
        A[p2] = tmp; 
    }
}

File: src/CPP/src/leetcode/SortColors.cpp
/*
 * [Source] https://leetcode.com/problems/sort-colors/
 * [Difficulty]: Medium
 * [Tag]: Array
 * [Tag]: Two Pointers
 * [Tag]: Sort
 */

#include <iostream>

using namespace std;

// [Solution]: Count the number and reconstruct the array
// [Corner Case]:
class Solution {
};

// [Solution]: Use two pointers point to the end of sorted part. Notice if swap 1, need first swap with the next 1 position, which is 2 already, so don't move pointer ahead.
/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/sortandsearch/SortColors.java
 */

/* Java solution
public class Solution {
    public void sortColors(int[] A) {
        int white=-1;
		int blue=-1;
		for(int i=0; i<A.length; i++){
			if(A[i]==0){
				white++;
				blue++;
				swap(A, white, i);
				if(blue==white) continue;
				swap(A, blue, i);
			}else if(A[i]==1){
				blue++;
				swap(A, blue, i);
			}
		}
    }
	
	public void swap(int[] A, int a, int b){
		int tmp=A[a];
		A[a]=A[b];
		A[b]=tmp;
    }
}
 */

int main() {
    Solution sol;

    return 0;
}
